documentation/gnmi/dash-gnmi-design.md

```diff
@@ -0,0 +1,133 @@
+# GNMI Interface for DASH
+
+## Overview
+
+SONiC should provide a gNMI server interface for DASH project.
+Below diagram shows the architecture of the DASH project, the host server and DPU card are connected through PCIE, and together they form a smart switch.
+
+![dash-arch](./images/gnmi-arch.svg)
+
+Both host server and DPU card are running SONiC image. GNMI server is on host server, orchagent and redis are on DPU card, GNMI server will use ZMQ over TCP to communicate with orchagent and redis.
+
+### Requirements
+
+* Set and get RPCs must be supported. Customers will use get RPC to retrieve DASH configurations and use set PRC to apply new DASH configurations.
+* Need to configure huge DASH table entries to APPL_DB, with high speed.
+* Minimal redis table scaling requirements:
+
+|Item |Expected value |
+|--|--|
+|VNETs|1024|
+|ENI|64 per card|
+|Outbound Routes per ENI|100K|
+|Inbound Routes per ENI|10K|
+|NSGs per ENI|10|
+|ACL rules per NSG|1000|
+|ACL prefixed per ENI|10*100K|
+|Max prefixed per rule|8K|
+|ACL ports per ENI|10*10K SRC/DST ports|
+|CA-PA mapping|10M per card|
+
+### Design considerations
+
+* For set RPC, GNMI calls the ZMQ interface, and ZMQ updates redis asynchronously.
+* For get RPC, GNMI reads from redis db directly.
+* The DASH table of APPL_DB would be encoded as protobuf to save memory consumption, and then GNMI needs to support protobuf encoding for DASH table.
+
+> 127.0.0.1:6379> hgetall "DASH_VNET_TABLE:vnet1"<br>
+> 1\) "pb"<br>
+> 2\) "\n\x010\x12$b6d54023-5d24-47de-ae94-8afe693dd1fc…"
+
+And proto message for DASH_VNET_TABLE is:
+
+> message Vnet {<br>
+> &ensp;string vni = 1;<br>
+> &ensp;string guid = 2;<br>
+> &ensp;repeated types.IpPrefix address_space = 3;<br>
+> &ensp;repeated string peer_list = 4;<br>
+> }
+
+* GNMI would not run Yang validation for protobuf encoding data.
+  * GNMI server can check ENI and VNET if necessary.
+
+# Design
+## Set RPC
+### Work flow
+![gnmi-set-flow](./images/gnmi-set-flow.svg)
+
+GNMI server would not run Yang validation, and it would invoke ZMQ interface to update.
```

Comment:
>would not run Yang validation [](http://example.com/codeflow?start=12&length=29)

if there is no validation inside gnmi, if gNMI client request has invalid payload, which component will detect error? Will the client finally get the gNMI error response?

Reply:
GNMI can't run yang validation for performance limitation. And GNMI can run some simple validation, if VNET or ENI does not exist, GNMI can return error response.